src/ser.rs

```diff
@@ -148,21 +158,44 @@ impl ser::SerializeMap for MapSerializer<'_> {
 
     fn serialize_key<T: ?Sized + Serialize>(&mut self, key: &T) -> Result<()> {
         debug_assert!(self.next_key.is_none());
-        self.next_key = Some(key.serialize(self.serializer)?);
+        let next_key = key.serialize(self.serializer)?;
+        if let MapResult::Object(_) = self.target {
+            if !next_key.is_string() {
+                return Err(Error::custom(
+                    "Map key is not a string and cannot be an object key",
+                ));
+            }
+        }
+
+        self.next_key = Some(next_key);
         Ok(())
     }
 
     fn serialize_value<T: ?Sized + Serialize>(&mut self, value: &T) -> Result<()> {
-        self.target.set(
-            &self.next_key.take().unwrap(),
-            &value.serialize(self.serializer)?,
-        );
+        match self.target {
+            MapResult::Map(ref map) => {
+                map.set(
+                    &self.next_key.take().unwrap(),
```

Comment:
Nitpick: can you please extract key & value calculation to variables before the `match`?